### PR TITLE
#11327 - Introduce bulk action 'Undo Deletion'

### LINKED ISSUE #11327: Introduce "Restore" bulk action [3]
### Problem Description

With #8465  the new grid for soft deleted entities was introduced for cases, contacts, events, event participants, immunizations, samples and travel entries. 

In this grid, bulk edit mode and bulk actions are possible. However, among the bulk actions there is currently 'Delete' but no 'Undo Deletion' bulk action, similar to how there is archive and de-archive.

### Proposed Change

Introduce the bulk action 'Undo deletion' in the grids that display soft deleted entities.

### Acceptance Criteria

- [x] An "Undo deletion" bulk action is shown in all directories specified above when filtering by deleted entities
- [x] Clicking the bulk action brings up a pop-up:
  - Caption: "Confirm undo deletion"
  - Content: "Are you sure you want to undo the deletion for all {selected_entry_count} selected {entity_name}s?"
  - Options: "No" and "Yes"
  - Should look similar to the pop-up that appears when e.g. bulk de-archiving cases
- [x] Grid is reloaded after confirming the action; entities for which the deletion has been undone no longer appear in the grid (since it should still be filtered by deleted entities)
- [x] Clicking the bulk action without selecting any entity brings up the usual warning message
- [x] Hide the "Delete" bulk action when filtering cases by deleted cases

### Additional Information
Make sure that the user is always allowed to undelete all the cases in the list (i.e. check whether there's any restriction in this regard for the archive/de-archive bulk action; if yes, we need to have something similar here, if no, we're probably fine).

### PR DESCRIPTION
**Pull request by sergiupacurariu**
Fixes #11327

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11802

File: sormas-api/src/main/resources/strings.properties [hunk 1]
```diff
@@ -985,14 +1000,22 @@ messageCaseIncidenceUnsupportedAgeGroup = Case incidence proportion can only be
 messageCaseReferredFromPoe = Case has been referred to the specified facility
 messageCaseRelationToEventWithoutDisease=It is not possible to link a case to an event if the disease of the event has not been set
 messageCaseSaved = Case saved
-messageCaseSavedClassificationChanged = Case saved. The classification was automatically changed to %s.
+messageCaseSavedClassificationChanged = Case saved. The c  lassification was automatically changed to %s.
```
Comment (MateStrysewske) [inline sormas-api/src/main/resources/strings.properties:1003]: Revert this change

File: sormas-api/src/main/resources/strings.properties [hunk 2]
```diff
@@ -985,14 +1000,22 @@ messageCaseIncidenceUnsupportedAgeGroup = Case incidence proportion can only be
...
 messageCasesMerged = Cases merged and duplicate case deleted.
+messageCasesRestored = All selected cases have been restored
+messageContactsRestored = All selected contacts have been restored
+messageEventsRestored = All selected events have been restored
+messageEventParticipantsRestored = All selected event participants have been restored
+messageImmunizationsRestored = All selected immunizations have been restored
+messageSamplesRestored = All selected samples have been restored
+messageTravelEntriesRestored = All selected travel entries have been restored
 messageCasesNotDeletedReasonExternalSurveillanceTool = Some cases were not deleted because the communication with the reporting tool failed.
+messageCasesNotRestored = The cases mentioned in the above ids list could not be restored.
```
Comment (MateStrysewske) [inline sormas-api/src/main/resources/strings.properties:1018]:
```suggestion
messageCasesNotRestored = The cases mentioned in the above list of UUIDs could not be restored.
```

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import com.vaadin.shared.ui.ContentMode;
+import com.vaadin.ui.Notification;
+import com.vaadin.ui.Window;
+
+import de.symeda.sormas.api.DeletableFacade;
+import de.symeda.sormas.api.i18n.I18nProperties;
+import de.symeda.sormas.api.i18n.Strings;
+import de.symeda.sormas.api.utils.HtmlHelper;
+
+public class DeleteUndeleController<F extends DeletableFacade> {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java:16]:
```suggestion
public class DeleteUndeleteController<F extends DeletableFacade> {
```

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java [hunk 4]
```diff
@@ -232,4 +232,14 @@ public void deleteAllSelectedItems(Collection<TravelEntryIndexDto> selectedRows,
 				});
 		}
 	}
+
+	public void undeleteSelectedCases(Collection<? extends TravelEntryIndexDto> selectedRows, Runnable callback) {
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java:236]:
```suggestion
	public void undeleteSelectedTravelEntries(Collection<? extends TravelEntryIndexDto> selectedRows, Runnable callback) {
```

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+	public void undeleteSelectedItems(List<String> entityUuids, F entityFacade, CoreEntityUndeleteMessages messages, Runnable callback) {
+		if (entityUuids.size() == 0) {
+			new Notification(
+				I18nProperties.getString(messages.getHeadingNoSelection()),
+				I18nProperties.getString(messages.getMessageNoSelection()),
+				Notification.Type.WARNING_MESSAGE,
+				false).show(Page.getCurrent());
+		} else {
+			int countNotRestoredCases = 0;
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java:26]:
```suggestion
			int countNotRestoredEntities = 0;
```

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+			StringBuilder notRestoredCases = new StringBuilder();
```
Comment (MateStrysewske) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java:27]:
```suggestion
			StringBuilder notRestoredEntities = new StringBuilder();
```

**Review by MateStrysewske — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/resources/strings.properties (hunk 1 above)
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings.properties:1006]: reverted

File: sormas-api/src/main/resources/strings.properties (hunk 2 above)
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings.properties:1021]: updated

File: sormas-api/src/main/resources/strings.properties [hunk 7]
```diff
@@ -1062,6 +1085,13 @@ messageEventsDearchived = All selected events have been de-archived
 messageEventsDeleted = All selected events have been deleted
 messageEventsSentToSurvnet = All selected events have been sent to SurvNet
 messageCountCasesNotDeleted = %s cases not deleted. CasesIds not deleted: %s
+messageCountCasesNotRestored = %s cases not restored. CasesIds not restored: %s
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-api/src/main/resources/strings.properties:1093]: changed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java (hunk 3 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java:16]: updated

File: sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java (hunk 4 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/travelentry/TravelEntryController.java:236]: changed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+import de.symeda.sormas.api.DeletableFacade;
+import de.symeda.sormas.api.i18n.I18nProperties;
+import de.symeda.sormas.api.i18n.Strings;
+import de.symeda.sormas.api.utils.HtmlHelper;
+
+public class DeleteUndeleController<F extends DeletableFacade> {
+
+	public void undeleteSelectedItems(List<String> entityUuids, F entityFacade, CoreEntityUndeleteMessages messages, Runnable callback) {
+		if (entityUuids.size() == 0) {
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java:19]: changed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java (hunk 6 above)
Comment (sergiupacurariu) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/utils/DeleteUndeleController.java:27]: changed

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11802

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11802

**Review by MateStrysewske — CHANGES_REQUESTED** [review]
- The confirmation pop-up when undeleting entities does not appear; the undeletion is done immediately
- The success notification is only displayed for a very short amount of time when the user moves their mouse because of the chosen notification type (it's basically not readable); use the tray notification instead

**Comment by sergiupacurariu** [comment]
issues regarding pop-up confirmation and tray notification have been fixed

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11802

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11802

**Pull request merged**